### PR TITLE
bounding box for triangulation and cell iterators [WIP]

### PR DESCRIPTION
**Pull request by danshapero**
I'm working on a project where I'll need to compute bounding boxes for triangulations and individual cells. This PR adds a function to the GridTools namespace to compute the bounding box for an entire triangulation. This feature was mentioned in #5117 . I'll add bounding boxes for cell iterators soon too.

I wanted to make a PR now just to check something before I continue. In the case where the spatial dimension of a triangulation is greater than the topological dimension, you can have cases where, say, the z-coordinates of the two points of the bounding box are equal (see the test case I added when `dim < spacedim`). However, the constructor for BoundingBox expects that `x[k] < y[k]` for each `k`, where `x` and `y` are the endpoints of the box. I changed the constructor for BoundingBox to accept `x[k] <= y[k]`, but I wanted to check in before proceeding to make sure this doesn't conflict with any of the work @GivAlz is doing.

### CODE REVIEW

File: source/grid/grid_tools.cc [hunk 1]
```diff
@@ -364,6 +364,26 @@ namespace GridTools
...
 
+
+  template <int dim, int spacedim>
+  BoundingBox<spacedim>
+  compute_bounding_box(const Triangulation<dim, spacedim> &tria)
+  {
+    const std::vector<Point<spacedim> > &vertices = tria.get_vertices();
+
+    auto boundary_points = std::make_pair(vertices[0], vertices[0]);
+    for (const auto &x: vertices)
```
Comment (bangerth) [inline source/grid/grid_tools.cc]: This doesn't quite work because not all elements of the vertices array are actually in use and may contain invalid values. Take a look at the documentation of Triangulation::get_vertices ().

File: source/grid/tria.cc [hunk 2]
```diff
@@ -11758,7 +11758,6 @@ Triangulation<dim, spacedim>::get_used_vertices () const
 
 
 
-
```
Comment (bangerth) [inline source/grid/tria.cc]: This is unrelated - just drop this change.

File: tests/grid/grid_tools_bounding_box.cc [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#include "../tests.h"
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_tools.h>
```
Comment (bangerth) [inline tests/grid/grid_tools_bounding_box.cc]: Can you at a brief comment at the top that explains what the test checks?

File: include/deal.II/grid/grid_tools.h [hunk 4]
```diff
@@ -170,6 +171,12 @@ namespace GridTools
   template <int dim, typename T>
   double cell_measure (const T &, ...);
 
+  /**
+   * Compute the smallest box containing the entire triangulation.
+   */
+  template <int dim, int spacedim>
+  BoundingBox<spacedim> compute_bounding_box(const Triangulation<dim, spacedim> &triangulation);
```
Comment (davydden) [inline include/deal.II/grid/grid_tools.h]: FYI (maybe helps with implementation, or re-use it) https://www.dealii.org/developer/doxygen/deal.II/namespaceGridTools.html#a0374dd9eae73c3c2badaa479e8ba8d55

File: source/grid/grid_tools.cc [hunk 5]
```diff
@@ -364,6 +364,31 @@ namespace GridTools
   }
 
 
+
+  template <int dim, int spacedim>
+  BoundingBox<spacedim>
+  compute_bounding_box(const Triangulation<dim, spacedim> &tria)
```
Comment (davydden) [inline source/grid/grid_tools.cc]: what should be an expected behaviour for `p::d::Tria`? Maybe throw an error if `ExcNotImplemented()` if it's provided?

File: include/deal.II/grid/grid_tools.h (hunk 4 above)
Comment (bangerth) [reply include/deal.II/grid/grid_tools.h]: Oh, good memory, @davydden -- we already have this function if you just pass `[](...) { return true; }` to that function!

Comment (davydden) [reply include/deal.II/grid/grid_tools.h]: > Oh, good memory

that's because my colleague @vishalkenchan added it in https://github.com/dealii/dealii/commit/0cc970cc57bb5a1ce9ee031d582e66f61a4c9fad :smile:

Comment (danshapero) [reply include/deal.II/grid/grid_tools.h]: Oof how did I miss that?!

File: source/grid/grid_tools.cc (hunk 5 above)
Comment (danshapero) [reply source/grid/grid_tools.cc]: Good point! Should I modify the function that I'm adding to deal with distributed triangulations, or the previous function that accepts an arbitrary predicate? You could ask the same question about both.

File: include/deal.II/base/bounding_box.h [hunk 6]
```diff
@@ -108,7 +108,7 @@ BoundingBox<spacedim, Number>::BoundingBox (const std::pair<Point<spacedim,Numbe
 {
   //We check the Bounding Box is not degenerate
   for (unsigned int i=0; i<spacedim; ++i)
-    Assert (boundary_points.first[i] < boundary_points.second[i],
+    Assert (boundary_points.first[i] <= boundary_points.second[i],
```
Comment (GivAlz) [inline include/deal.II/base/bounding_box.h]: These days I've been working hard on my code an I lost this, sorry!!!

So, I have no problem with creating an empty bounding box like this one: the Assert is there because, as far as I'm concerned, if I'm creating an empty box "manually" by writing the single points then I'm doing something wrong, probably...
If you need to create one and then use compute_bounding_box then you could avoid altogether using the constructor with separate points; but if you have some good reason to allow building by hand an empty box than I have no problem...

**Comment by davydden** [comment]
@danshapero i suppose it's not a WIP anymore?

**Comment by danshapero** [comment]
@davydden Actually there's one more thing I wanted to change. The function for computing the bounding box of a cell operates on a `CellAccessor`, but I think it might be better if it worked on a `TriaIterator` instead. Unfortunately I run into a bunch of incomprehensible compile errors if I try to define it as taking an argument of type `typename Triangulation<dim, spacedim>::cell_iterator&`. I also wonder if this function might be better implemented as a method of `TriaIterator` rather than a free function.

In any case, I'll fix that and and let you know when I'm ready for another look.

**Comment by bangerth** [comment]
Yes, make that a member function of `TriaAccessor`. It already has a bunch of functions that return geometric properties (such as `center()`, `measure()`, etc). It would fit well there.

**Comment by danshapero** [comment]
I added `bounding_box` as a member of `TriaAccessor` and rebased it down to 2 commits. Ready to go on my end!

File: source/grid/grid_tools.cc (hunk 5 above)
Comment (davydden) [reply source/grid/grid_tools.cc]: Maybe this one only as to me it suggest the box and f the whole thing, not only of locally owned DoFs. Or maybe just mention it in the docu as a warning.

File: source/grid/grid_tools.cc [hunk 7]
```diff
@@ -364,6 +364,43 @@ namespace GridTools
...
+    };
+
+    return BoundingBox<spacedim>(compute_bounding_box(tria, predicate));
+  }
+
+
+
+  template <int dim, int spacedim>
+  BoundingBox<spacedim>
+  compute_bounding_box(const CellAccessor<dim, spacedim> &accessor)
```
Comment (bangerth) [inline source/grid/grid_tools.cc]: Why do you still need this? Can't you just call the function you implemented in `TriaAccessor`?

File: source/grid/grid_tools.cc [hunk 8]
```diff
@@ -364,6 +364,43 @@ namespace GridTools
...
+  BoundingBox<spacedim>
+  compute_bounding_box(const Triangulation<dim, spacedim> &tria)
+  {
+    using iterator = typename Triangulation<dim, spacedim>::active_cell_iterator;
+    const auto predicate = [](const iterator &)
+    {
+      return true;
+    };
+
+    return BoundingBox<spacedim>(compute_bounding_box(tria, predicate));
```
Comment (bangerth) [inline source/grid/grid_tools.cc]: No need for the explicit cast to `BoundingBox` (though also not wrong).

File: source/grid/grid_tools.cc (hunk 5 above)
Comment (davydden) [reply source/grid/grid_tools.cc]: Ps. The function you re-use, in my interpretation, does/should not promise to do the union of the results over MPI communicator

File: source/grid/grid_tools.cc [hunk 9]
```diff
@@ -364,6 +364,22 @@ namespace GridTools
...
+  BoundingBox<spacedim>
+  compute_bounding_box(const Triangulation<dim, spacedim> &tria)
+  {
+    using iterator = typename Triangulation<dim, spacedim>::active_cell_iterator;
+    const auto predicate = [](const iterator &)
+    {
+      return true;
+    };
+
+    return compute_bounding_box(tria, predicate);
```
Comment (davydden) [inline source/grid/grid_tools.cc]: maybe before returning, in case of `p::d::Tria`, use `Utilities::MPI::min/max` to do union of boxes? Or add assert that `p::d::Tria` is not supported.

Comment (davydden) [reply source/grid/grid_tools.cc]: @bangerth what do you think about `compute_bounding_box` of p::d::Tria?

Comment (bangerth) [reply source/grid/grid_tools.cc]: I think I'm ok with the current status. One could add a sentence such as "If the triangulation passed as an argument to this function is a parallel::distributed::Triangulation, then each processor will compute a bounding box that encloses all locally owned, ghost, and artificial cells. In the case of a domain without curved boundaries, these bounding boxes will all agree between processors because the union of areas occupied by artificial and ghost cells equals the union of the areas occupied by the cells owned by other processors. However, if the domain has curved boundaries, that is not the case and the bounding box returned may only be appropriate for the current processor, but may be different from the bounding boxes computed on other processors."

Alternatively, as you suggest, one could take the union over all processors to get a single, globally unique, bounding box.

**Comment by danshapero** [comment]
There seems to be a failure on gcc-4.8.4, possibly having something to do with lambdas? I'll see if I can reproduce and try to find a workaround.

**Comment by bangerth** [comment]
You should be fine if you explicitly convert the lambda to the required `std::function` type at the calling site (line 378).

File: source/grid/grid_tools.cc (hunk 9 above)
Comment (davydden) [reply source/grid/grid_tools.cc]: Oh yes, you are right! We are looking at active cells in predicate, not locally owned.

**Comment by davydden** [comment]
@danshapero looks like you need to rename your tests to be `_03`.

**Comment by davydden** [comment]
@danshapero there's a minor indentation issue, you can fix it manually or run `./contrib/utilities/indent`

**Comment by danshapero** [comment]
Alright, now that I've cast the lambda to a `std::function` it no longer breaks on gcc-4.8. Unless there's something more you'd like changed I think it's good to go!

File: source/grid/tria_accessor.cc [hunk 10]
```diff
@@ -1110,6 +1110,29 @@ measure () const
...
 
+template <int structdim, int dim, int spacedim>
+BoundingBox<spacedim>
+TriaAccessor<structdim, dim, spacedim>::
+bounding_box () const
+{
+  std::pair<Point<spacedim>, Point<spacedim> > boundary_points =
+    std::make_pair(this->vertex(0), this->vertex(0));
+
+  for (unsigned int v=0; v<GeometryInfo<structdim>::vertices_per_cell; ++v)
```
Comment (davydden) [inline source/grid/tria_accessor.cc]: you can start from one here.

Comment (danshapero) [reply source/grid/tria_accessor.cc]: Every cycle counts ;)

**Comment by bangerth** [comment]
Thanks for sticking with us through the cycles of review!
